### PR TITLE
gh-2476 Stop defaulting access roles to empty arrays. 

### PR DESCRIPTION
**Pull request by m29827**
Read and Write access role properties are mutually exclusive with read and write AccessPredicate properties.
Also added some @Test annotations (see #2452) and corrected the tests which were broken and now executing again.

### CODE REVIEW

File: core/operation/src/main/java/uk/gov/gchq/gaffer/named/operation/AddNamedOperation.java [hunk 1]
```diff
@@ -51,7 +52,7 @@
  * A {@code AddNamedOperation} is an {@link Operation} for creating a new {@link NamedOperation}
  * and adding it to a Gaffer graph.
  */
-@JsonPropertyOrder(value = {"class", "operationName", "description", "score", "operations"}, alphabetic = true)
+@JsonPropertyOrder(value = {"class", "operationName", "description", "score", "labels", "operationChain", "input", "overwriteFlag", "parameters", "readAccessRoles", "writeAccessRoles", "readAccessPredicate", "writeAccessPredicate"}, alphabetic = true)
```
Comment (GCHQDev404) [inline core/operation/src/main/java/uk/gov/gchq/gaffer/named/operation/AddNamedOperation.java:55]: Was this ordering required anything, is it neater than remaining being alphabetical order?

File: core/operation/src/main/java/uk/gov/gchq/gaffer/named/operation/AddNamedOperation.java [hunk 2]
```diff
@@ -173,8 +174,8 @@ public AddNamedOperation shallowClone() {
                 .name(operationName)
                 .labels(labels)
                 .description(description)
-                .readAccessRoles(readAccessRoles.toArray(new String[readAccessRoles.size()]))
-                .writeAccessRoles(writeAccessRoles.toArray(new String[writeAccessRoles.size()]))
+                .readAccessRoles(readAccessRoles == null ? null : readAccessRoles.toArray(new String[readAccessRoles.size()]))
```
Comment (GCHQDev404) [inline core/operation/src/main/java/uk/gov/gchq/gaffer/named/operation/AddNamedOperation.java:177]:
```suggestion
                .readAccessRoles(isNull(readAccessRoles) ? null : readAccessRoles.toArray(new String[readAccessRoles.size()]))
```

File: core/operation/src/main/java/uk/gov/gchq/gaffer/named/operation/AddNamedOperation.java [hunk 3]
```diff
@@ -173,8 +174,8 @@ public AddNamedOperation shallowClone() {
...
+                .writeAccessRoles(writeAccessRoles == null ? null : writeAccessRoles.toArray(new String[writeAccessRoles.size()]))
```
Comment (GCHQDev404) [inline core/operation/src/main/java/uk/gov/gchq/gaffer/named/operation/AddNamedOperation.java:178]:
```suggestion
                .writeAccessRoles(isNull(writeAccessRoles) ? null : writeAccessRoles.toArray(new String[writeAccessRoles.size()]))
```

File: core/operation/src/main/java/uk/gov/gchq/gaffer/named/operation/AddNamedOperation.java [hunk 4]
```diff
@@ -301,12 +302,24 @@ public Builder description(final String description) {
         }
 
         public Builder readAccessRoles(final String... roles) {
-            Collections.addAll(_getOp().getReadAccessRoles(), roles);
+            if (null == roles) {
```
Comment (GCHQDev404) [inline core/operation/src/main/java/uk/gov/gchq/gaffer/named/operation/AddNamedOperation.java:305]:
```suggestion
            if (isNull(roles)) {
```

File: core/operation/src/main/java/uk/gov/gchq/gaffer/named/operation/AddNamedOperation.java [hunk 5]
```diff
@@ -301,12 +302,24 @@ public Builder description(final String description) {
...
+                _getOp().setReadAccessRoles(null);
+            } else if (null == _getOp().getReadAccessRoles()) {
```
Comment (GCHQDev404) [inline core/operation/src/main/java/uk/gov/gchq/gaffer/named/operation/AddNamedOperation.java:307]:
```suggestion
            } else if (isNul( _getOp().getReadAccessRoles())) {
```

File: core/operation/src/main/java/uk/gov/gchq/gaffer/named/operation/AddNamedOperation.java [hunk 6]
```diff
@@ -301,12 +302,24 @@ public Builder description(final String description) {
...
+                _getOp().setReadAccessRoles(Arrays.asList(roles));
+            } else {
+                Collections.addAll(_getOp().getReadAccessRoles(), roles);
+            }
             return _self();
         }
 
         public Builder writeAccessRoles(final String... roles) {
-            Collections.addAll(_getOp().getWriteAccessRoles(), roles);
+            if (null == roles) {
```
Comment (GCHQDev404) [inline core/operation/src/main/java/uk/gov/gchq/gaffer/named/operation/AddNamedOperation.java:316]:
```suggestion
            if (isNull(roles)) {
```

File: core/operation/src/main/java/uk/gov/gchq/gaffer/named/operation/AddNamedOperation.java [hunk 7]
```diff
@@ -301,12 +302,24 @@ public Builder description(final String description) {
...
+                _getOp().setWriteAccessRoles(null);
+            } else if (null == _getOp().getWriteAccessRoles()) {
```
Comment (GCHQDev404) [inline core/operation/src/main/java/uk/gov/gchq/gaffer/named/operation/AddNamedOperation.java:318]:
```suggestion
            } else if (isNull(_getOp().getWriteAccessRoles())) {
```

File: core/operation/src/test/java/uk/gov/gchq/gaffer/named/operation/AddNamedOperationTest.java [hunk 8]
```diff
@@ -94,20 +96,16 @@ public void shouldJsonSerialiseAndDeserialise() {
...
                 "\"writeAccessPredicate\" : {%n" +
-                "    \"class\" : \"uk.gov.gchq.gaffer.access.predicate.CustomAccessPredicate\",%n" +
-                "    \"userId\" : \"CreatingUserId\",%n" +
-                "    \"map\" : {%n" +
-                "        \"WriteKey\": \"WriteValue\"%n" +
-                "    },%n" +
-                "    \"auths\" : [ \"CustomWriteAuth1\", \"CustomWriteAuth2\" ]%n" +
+                "    \"class\" : \"uk.gov.gchq.gaffer.access.predicate.AccessPredicate\",%n" +
+                "    \"userPredicate\" : {%n" +
+                "       \"class\" : \"uk.gov.gchq.gaffer.access.predicate.user.CustomUserPredicate\"%n" +
```
Comment (GCHQDev404) [inline core/operation/src/test/java/uk/gov/gchq/gaffer/named/operation/AddNamedOperationTest.java:107]: Where has the predicate gone that is being tested?
```CustomWriteAuth1```

Comment (m29827) [reply core/operation/src/test/java/uk/gov/gchq/gaffer/named/operation/AddNamedOperationTest.java:107]: This test was not being executed, it's another example of Junit 5 not discovering and running overridden `@Test` methods.  When the `@Test` annotation was added I discovered it was broken. I've removed the `userid`, `map` and `auths` fields as they no longer exist in `uk.gov.gchq.gaffer.access.predicate.CustomAccessPredicate`.

**Review by t92549 — APPROVED** [review]

**Pull request merged**